### ISSUE TITLE
`Polygon has invalid geometry` errors when downloading building footprints

### ISSUE THREAD

**Issue by 00krishna** [gboeing/osmnx#279]
#### Problem description (what did you do, what did you expect to happen, and what actually happened)

I am using the `ox.footprints.footprints_from_polygon()` function to download the building footprints from Tanzania. I pulled the Shapely polygon for the Tanzanian state of Dar-es-Salaam and put that into the function `osmnx` function. 

Here is a snippet of the MULTIPOLYGON:

```
MULTIPOLYGON (((39.49763870239263 -7.187915802001953, 39.49763870239263 -7.189027786254883, 39.49791717529303 -7.189027786254883, 39.49791717529303 -7.189303874969482, 39.49819564819342 -7.189303874969482, 39.49819564819342 -7.189027786254883, 39.49847030639677 -7.189027786254883, 39.49847412109375 -7.18874979019165, 39.49902725219755 -7.18874979019165, 39.49902725219755 -7.188471794128418, 39.50180435180664 -7.188471794128418, 39.50180435180664 -7.18874979019165, 39.50291824340837 -7.18874979019165, 39.50291824340837 -7.189303874969482, 39.50319290161156 -7.189305782318058, 39.50319290161156 -7.189859867095947, 39.50347137451195 -7.189859867095947, 39.50347137451195 -7.19013786315918, 39.50374984741217 -7.19013786315918, 39.50374984741217 -7.190972805023193, 39.50152587890625 -7.190972805023193, 39.50152587890625 -7.190693855285645, 39.50069427490229 -7.190693855285645, 39.50069427490229 -7.190415859222412, 39.5004158020019 -7.190415859222412, 39.5004158020019 -7.190693855285645, 39.50013732910173 -7.190693855285645, 39.50013732910173 -7.190972805023193, 39.49986267089849 -7.190972805023193, 39.49986267089849 -7.191250801086426, 39.49930572509794 -7.191250801086426, 39.49930572509794 -7.190972805023193, 39.49874877929716 -7.190972805023193, 39.49874877929716 -7.190693855285645, 39.49847412109375 .... 
```
dar2 = tanz_data.loc[tanz_data.ID_1==2,'geometry'].unary_union
gdf = ox.footprints.footprints_from_polygon(dar2, retain_invalid=False)

The function will start running and download a bunch of footprints. But then towards the end, I will get errors like the one below. I can see these errors in the terminal associated with the jupyter notebook.

```
Polygon has invalid geometry: [3200697299, 3200697323]
Polygon has invalid geometry: [3441252415, 3713810622]
Polygon has invalid geometry: [3683960900, 3683960912]
Polygon has invalid geometry: [338935148, 3698315925]
Polygon has invalid geometry: [3698437744, 3698376185]
Polygon has invalid geometry: [3698437744, 3742824730]
Polygon has invalid geometry: [3698376185, 3709146540]
Polygon has invalid geometry: [252499785, 282762863]
Polygon has invalid geometry: [3698350767, 3441252415]
Polygon has invalid geometry: [252499818, 253098908]
Polygon has invalid geometry: [1500081793, 1500081763]
```
The error I see in the notebook has a traceback like:

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-68-2a4588788aab> in <module>
----> 1 gdf = ox.footprints.footprints_from_polygon(dar2, retain_invalid=False)
      2 gdf_proj = ox.project_gdf(gdf)

/media/hayagriva/anaconda3/envs/ox/lib/python3.7/site-packages/osmnx/footprints.py in footprints_from_polygon(polygon, footprint_type, retain_invalid)
    328 
    329     return create_footprints_gdf(polygon=polygon, footprint_type=footprint_type, 
--> 330                                  retain_invalid=retain_invalid)
    331 
    332 

/media/hayagriva/anaconda3/envs/ox/lib/python3.7/site-packages/osmnx/footprints.py in create_footprints_gdf(polygon, north, south, east, west, footprint_type, retain_invalid)
    240     # remove supporting geometry from footprints dictionary
    241     for item in pop_list:
--> 242         footprints.pop(item)
    243 
    244     gdf = gpd.GeoDataFrame(footprints).T

KeyError: 453498330
```

#### What operating system, architecture, Python version, and OSMnx version are you using?
OSMnx version 0.90, running on Ubuntu Linux 18.04 x64. Python version 3.7.1. I installed the package with conda as per the instructions in the issues. 

#### Complete list of your environment's packages and their versions (for example, run `conda list` or `pip list` then paste the output below, between the two "details" tags)

<details>
# Paste the output of your Python packages and their versions here, between these two "details" tags
```
➜ pip list
Package            Version
------------------ -----------
altair             3.0.0
asn1crypto         0.24.0
attrs              19.1.0
backcall           0.1.0
bleach             3.1.0
branca             0.3.1
certifi            2019.3.9
cffi               1.12.3
chardet            3.0.4
Click              7.0
click-plugins      1.1.1
cligj              0.5.0
cryptography       2.5
cycler             0.10.0
decorator          4.4.0
defusedxml         0.6.0
descartes          1.1.0
entrypoints        0.3
Fiona              1.8.6
folium             0.8.3
GDAL               2.4.0
geographiclib      1.49
geopandas          0.5.0
geopy              1.19.0
idna               2.8
ipykernel          5.1.0
ipython            7.5.0
ipython-genutils   0.2.0
ipywidgets         7.4.2
jedi               0.13.3
Jinja2             2.10.1
jsonschema         3.0.1
jupyter            1.0.0
jupyter-client     5.2.4
jupyter-console    6.0.0
jupyter-core       4.4.0
kiwisolver         1.1.0
mapclassify        2.0.1
MarkupSafe         1.1.1
matplotlib         3.0.3
mistune            0.8.4
munch              2.3.2
nbconvert          5.4.1
nbformat           4.4.0
networkx           2.3
notebook           5.7.8
numpy              1.16.3
osmnx              0.9
pandas             0.24.2
pandocfilters      1.4.2
parso              0.4.0
pexpect            4.7.0
pickleshare        0.7.5
pip                19.1
prometheus-client  0.6.0
prompt-toolkit     2.0.9
ptyprocess         0.6.0
pycparser          2.19
Pygments           2.3.1
pyOpenSSL          19.0.0
pyparsing          2.4.0
pyproj             1.9.6
pyrsistent         0.15.1
PySocks            1.6.8
python-dateutil    2.8.0
pytz               2019.1
pyzmq              18.0.0
qtconsole          4.4.3
requests           2.21.0
Rtree              0.8.3
scipy              1.2.1
Send2Trash         1.5.0
setuptools         41.0.1
Shapely            1.6.4.post2
six                1.12.0
terminado          0.8.2
testpath           0.4.2
toolz              0.9.0
tornado            6.0.2
traitlets          4.3.2
urllib3            1.24.2
vincent            0.4.4
wcwidth            0.1.7
webencodings       0.5.1
wheel              0.33.1
widgetsnbextension 3.4.2
```
</details>

#### Code snippet that independently and completely reproduces the issue

```python
```
Code inserted above.

**Comment by AtelierLibre** [comment]
I had a quick look at this and I have recreated your error using a boundary that I downloaded from Overpass-Turbo but it's not immediately obvious to me what is going on here.

Working with a smaller area that contains the way (https://www.openstreetmap.org/?way=453498330) referenced in the key error works:

`ox.footprints_from_point((-6.87851,39.17298), 100).plot()`

I would need more time to look into this further.

**Comment by gboeing** [comment]
Thanks @AtelierLibre. Any new insights?

**Comment by AtelierLibre** [comment]
@gboeing I don't, but I'll have a look at it as I'm working through the other one.

**Comment by 00krishna** [comment]
Thanks for your help on this guys. Seems like if there is a single invalid or old geometry in an area, then that will kill the download or the display for the entire plot or image, etc. My friend tried to get around this by basically downloading and then bisecting areas with errors and then downloading the bisections, until he could limit the areas with problem geometries. But that is not a particularly quick or elegant solution.

**Comment by AtelierLibre** [comment]
I think there may be a couple of things going on here: the key error and the invalid geometry. #283 should catch the key error. I'll look at the invalid geometry as part of #270 and #278.

Your original download area is quite large (approx. 100km x 68km). This is split into multiple responses by `osm_footprints_download`. I suspect this creates some duplicate footprints along the edges of the responses. If they are supporting geometry `create_footprints_gdf` would attempt to delete them from the final footprints dictionary multiple times and generate a key error. #283 should catch that and print a list of potential duplicates it can't find.

**Comment by 00krishna** [comment]
@AtelierLibre I was working with a large drone image that is about 30,000 x 30,000 pixels in size for Dar Es Salaam. I just pulled the extent of the image and then fed that to OSMNX. 

Just to give a little background, I had to do it this way because just referencing Dar Es Salaam as a location--like in the tutorial--was returning a single point. Seems like Dar Es Salaam is both a city point and a district, etc., in the Nominatrim database. So I had to go the route of getting the bounding box around the image and then submitting that to OSMNX. 

Do you mean that if I used smaller spatial extents, then that would work okay? Or I think what you are saying is that in generating the multiple responses from the server, the duplicate footprints along the edges were creating the errors.

**Comment by AtelierLibre** [comment]
@00krishna While this isn't fixed, using a spatial extent smaller than 50km x 50km would avoid the multiple queries and hence the duplicates and the key error. I have filtered the approx. 1,000,000 ways that your query returns to show just the duplicates. You can see in the image below how they align along the edges of the sub-queries. Fixing/avoiding the key error should be relatively straightforward.

However, at the moment you are quite likely to run into the separate issue that relations that contain open ways are creating invalid geometries. Sometimes these are correctly drawn buildings, sometimes they are tagging mistakes. You can also see an example of this in the image. The curving line is a bus route that has been tagged as a building [https://www.openstreetmap.org/relation/5440611](https://www.openstreetmap.org/relation/5440611) - for some reason there are quite a few of these in Dar Es Salaam.

I am looking at both of these issues but the second one in particular is going to take a bit of time.

![Screenshot from 2019-05-11 00-19-55](https://user-images.githubusercontent.com/36126654/57562086-cdada800-7387-11e9-9c78-d7c89f2326d1.png)

**Comment by 00krishna** [comment]
Okay I understand now. So there are two issues. One is that there are some invalid geometries depending on the location, etc. This is the bus example you provide. These are cases where say a geometry is mislabeled, or is incomplete. 

The second problem is that for queries larger than 50km x 50km, the queries will overlap and generate duplicates which creates the key error.

So I was running into both of these at the same time :). Luck of the draw I guess. So a workaround for problem 2--overlapping queries, I can just take smaller tiles of my larger extent. For the first problem with really invalid geometries, that is going to be harder to do. 

Hmm, is there anything that I could do to help out? I am happy to help out, though I don't really know the inner workings of the OSMNx package yet. Or if there is a way to break this into smaller problems I could take some of that on, if it would help. LMK.

**Comment by gboeing** [comment]
resolved in #289
